tools/generate.py

```diff
@@ -1,276 +1,404 @@
+
+        All results are gradually saved into database tables in
+        order to mitigate the memory footprint of the process.
+
+        Parameters
+        ----------
+        sentences_detailed_path : Path
+            path of the 'sentences_detailed.csv' weekly dump file
+        sentence_blacklist : list
+            the integer ids of the sentences that are not taken
+            into account for this counting
+        buffer_length : int, optional
+            the maximum number of n-grams for which counts are kept in RAM,
+            by default 5,000,000.
+            Note that process speed mainly depends on the amount of data
+            written to disk and consequently increases with the buffer size.
+        """
+        # the former database is overwritten
+        if self._fp.exists():
+            print(f"Deleting former {self._fp.name} database")
+            self._fp.unlink()
+        # initialize tables
+        self._init_db()
+
+        user_lang_score = defaultdict(int)
+        for n in range(MAX_NGRAM_SIZE, MIN_NGRAM_SIZE - 1, -1):
+            table_name = f"grams{n}"
+            lang_ngram_cnt = defaultdict(lambda: defaultdict(int))
+            with open(sentences_detailed_path, "r", encoding="utf-8") as f:
+                for line_id, line in enumerate(f):
+                    self._print_status(line_id, ngram_size=n)
+                    try:
+                        fields = line.rstrip().split("\t")
+                        sent_id, lang, text, user = fields[:4]
+                    except IndexError:
+                        print(f"Skipped erroneous line {line_id}: {line}")
+                        continue
+
+                    # sentences with an unset language are ignored
+                    if lang == "\\N" or lang == "":
+                        continue
+                    # sentences with an id matching the blacklist are ignored
+                    if int(sent_id) in sentence_blacklist:
+                        continue
+
+                    # compute user contribution score during last file reading
+                    if n == MIN_NGRAM_SIZE:
+                        user_lang_score[(user, lang)] += len(text)
+
+                    # increment hit counts for each ngram in the sentence
+                    for i in range(len(text) - n + 1):
+                        ngram = text[i : i + n]
+                        lang_ngram_cnt[lang][ngram] += 1
+
+                    # when buffer is full, save it into table and empty it
+                    tot_ngrams = sum(len(v) for v in lang_ngram_cnt.values())
+                    if tot_ngrams >= buffer_length:
+                        for lang, ngram_hits in lang_ngram_cnt.items():
+                            self._upsert_ngram_hits(
+                                ngram_hits, lang, table_name
                             )
-            conn.commit()
+                        lang_ngram_cnt = defaultdict(lambda: defaultdict(int))
```

Comment:
Maybe use lang_ngram_cnt.clear() instead to avoid duplicating the dictionary structure definition.